### PR TITLE
refactor: using C++11/14/17 features in dash specifict code

### PR DESCRIPTION
**Pull request by knst**
## Issue being fixed or feature implemented
This refactoring helps to make code more specific and clear.
There's using syntax feature from modern C++ such as 'enum class', structure bindings in loops, declaration variables inside if/switch statements, etc.

## What was done?
This PR is based on @PastaPastaPasta 's PR https://github.com/dashpay/dash/pull/4472 

There excluded changes related to using std::optional. Let's decide firstly about `Result` class: https://github.com/dashpay/dash/pull/5109 

## How Has This Been Tested?
Run unit/functional tests

## Breaking Changes
No breaking changes

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

**Comment by PastaPastaPasta** [comment]
(Also, please address CI failures)

**Comment by knst** [comment]
```
commit 417612644014c26d2c679eaeabb5af6dba448643
    fix: return size_t without conversion to int
```
This changes are dropped, because confusion about size_t type in pushKV in univalue.h

Should be done after this backport:
     Merge bitcoin/bitcoin#25611: univalue: Avoid brittle, narrowing and verbose integral type confusions

File: src/coinjoin/client.cpp [hunk 1]
```diff
@@ -1363,9 +1363,9 @@ bool CCoinJoinClientSession::PrepareDenominate(int nMinRounds, int nMaxRounds, s
         return true;
     }
 
-    for (const auto& pair : vecPSInOutPairsRet) {
-        mixingWallet.LockCoin(pair.first.prevout);
-        vecOutPointLocked.push_back(pair.first.prevout);
+    for (const auto& inout : vecPSInOutPairsRet) {
+        mixingWallet.LockCoin(inout.first.prevout);
+        vecOutPointLocked.push_back(inout.first.prevout);
```
Comment (UdjinM6) [inline src/coinjoin/client.cpp:1368]:
```suggestion
    for (const auto& [txDsIn, _] : vecPSInOutPairsRet) {
        mixingWallet.LockCoin(txDsIn.prevout);
        vecOutPointLocked.push_back(txDsIn.prevout);
```

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
one suggestion, LGTM otherwise

File: src/coinjoin/client.cpp (hunk 1 above)
Comment (knst) [reply src/coinjoin/client.cpp:1368]: `_` is not used variable, and [[unused]] is not allowed to be used inside structure binding.
I want to keep this case as it is.

Comment (UdjinM6) [reply src/coinjoin/client.cpp:1368]: We do this in quite a few places already. Why would this be an issue?

Comment (PastaPastaPasta) [reply src/coinjoin/client.cpp:1368]: Agree with Udjin here

Comment (knst) [reply src/coinjoin/client.cpp:1368]:
```
strAutoDenomResult = _("Session timed out.");                                  
```
`_` is used in this file and variable with name "_" will shadow it.

Comment (UdjinM6) [reply src/coinjoin/client.cpp]: Hmm.. That's a valid point I guess 👍

**Review by UdjinM6 — APPROVED** [review]
utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for squash merge

**Pull request merged**